### ISSUE TITLE
2.4 beta colates differently than 2.3.6 on darwin/arm64

### ISSUE THREAD

**Issue by idcmp** [mixxxdj/mixxx#12517]
### Bug Description

On 2.3.4(and .6) crates are sorted such that "!" comes before ":". On the 2.4.0 beta the ":" crates come before the "!" ones. 

I'm guessing playlists do the same.

### Version

2.4.0

### OS

macOS/arm64

**Comment by idcmp** [comment]
(regression, 2.4)

**Comment by ronso0** [comment]
You're referring to the sidabar I assume?
I can not confirm on Ubuntu with Qt 5.12.8, sort order is identical in 2.3.x and 2.4
The crates and playlist DAOs use `mixxx::DbConnection::collateLexicographically` to collect respective items for the sidebar, and AFAICT dbconnection didn't change significantly.

The cause must be elsewhere. Which Qt versions are 2.3 and 2.4 built for you?
For the track menu the sort order was adjusted to match that in the sidebar #11023

**Comment by idcmp** [comment]
> You're referring to the sidabar I assume?

If that has the preview deck+search bar ontop and option cover art at the bottom, then yup!

I'm not 100% sure how to check, so for 2.3.6:

```
$ strings /Applications/mixxx.app/Contents/Frameworks/QtCore.framework/QtCore | grep -E '^Qt 5'
Qt 5.12.3 (x86_64-little_endian-lp64 shared (dynamic) release build; by Clang 9.1.0 (clang-902.0.39.2) (Apple))
```

In 2.4, it looks like they're not used as frameworks, so:
````
$ strings mixxx | grep -E '^Qt 5'
Qt 5.12.5
Qt 5.12.5 (arm64-little_endian-lp64 static release build; by Clang 12.0.0 (clang-1200.0.32.29) (Apple))
idcmpbook-4:MacOS idcmp$ pwd
/Volumes/mixxx-2.4-beta-328-gf610d58a25/mixxx.app/Contents/MacOS
````

HTH!

**Comment by daschuer** [comment]
collateLexicographically dose not work with 2.3.x on macOS, because the qt uses a different libsqlite3 than linked directly to Mixxx. 
This has been fixed in 2.4 by building everything from the same source using vcpkg.

So this is no bug, but a feature. 

Maybe we could reference this explicit in the changelog.

**Comment by idcmp** [comment]
Got it. Thank you!

**Comment by ronso0** [comment]
@daschuer You left this open as reminder? I thought, for collecting items for the changelog, we go through the closed issues/PRs with the `Changelog` label, therefore I'd close this. Wdyt?

**Comment by ywwg** [comment]
+1 to closing this

**Comment by idcmp** [comment]
When writing the changelog entry please offer a tip of what the collation order is, I can't get anything else to sort the way Mixxx 2.4 is sorting. Both in ASCII and UNICODE the order feels random:

* `sort` will do: !, %, +, -, :, @, A-Z, ~.  -- (I tried a few different `LC_COLLATE` and `LANG` options -- this is ASCII ordering)
* Mixx 2.3.6 will do: !, %, +, -, :, @, A-Z, ~ (what `sort` does).

* Mixxx 2.4 wants to do: -, :, !, @, %, +, ~, A-Z.

**Comment by ronso0** [comment]
Tbh the 2.4 sorting looks best to me, except that + and -

> I can't get anything else to sort the way Mixxx 2.4 is sorting

How did you test? Would require other Qt6 apps to check whether it can be improved.

**Comment by idcmp** [comment]
Okay, I went down a bit of a rabbit hole here.

To start, I made a simple text file, numbers in brackets are decimal from `man 7 ascii`:

```
$ cat testing
A aaah (65)
Z zzz (90)
% percent (37)
: colon (58)
@ at (64)
! exclaim (33)
~ twiddle (126)
- dash (45)
+ plus (43)

```
Then I asked `sort` (I'm on macOS):

```
$ cat testing | sort

! exclaim (33)
% percent (37)
+ plus (43)
- dash (45)
: colon (58)
@ at (64)
A aaah (65)
Z zzz (90)
~ twiddle (126)
```

I tried various `LANG=`/ `LC_ALL=` / `LC_COLLATE=`, but I'm very much out of my depth here and they all sorted similarly. 

I was able to find https://onlinetexttools.com/sort-text-lines which sorts Mixxx 2.4 style:

```

- dash (45)
: colon (58)
! exclaim (33)
@ at (64)
% percent (37)
+ plus (43)
~ twiddle (126)
A aaah (65)
Z zzz (90)
```

The site claims to be sorting in "ASCII order", but it's not.  I don't know Python that well, but I did try this:
```
$ python3
>>> stuff = [ "! exclaim", "% precent", "+ plus", "- dash", ": colon", "@ at", "A aahh", "Z zzz", "~ twiddle" ]
>>> sorted(stuff)
['! exclaim', '% precent', '+ plus', '- dash', ': colon', '@ at', 'A aahh', 'Z zzz', '~ twiddle']
>>>
```

Then I installed `pyuca` - the unicode collator:

```
$ pip3 install pyuca
$ python3
>>> import pyuca
>>> collator = pyuca.Collator()
>>> sorted(stuff,key=collator.sort_key)
['- dash', ': colon', '! exclaim', '@ at', '% precent', '+ plus', '~ twiddle', 'A aahh', 'Z zzz']
>>>
```

So it looks like the answer is that it uses the Unicode Collation Algorithm (UCA) which various other parts of macOS (and Linux) do not by default. 

So to sum up, "sidebar entries are now sorted using the Unicode Collation Algorithm (UCA), and not in ASCII order."

**Comment by daschuer** [comment]
Thank you for investigations. I will us that sentences in the changelog.

**Comment by daschuer** [comment]
https://github.com/mixxxdj/mixxx/pull/12570/commits/00e7cece370f85a3c305b4a33023b831fa927301
